Add mimalloc submodule, and default to mimalloc where possible

nwc10 opened the pull request:
Permit building with mimalloc instead of libc malloc

Implement this by statically include mimalloc as a single source. Default to using mimalloc if `Configure.pl's` probe detects that the compiler provides C11 atomics, otherwise continue to use the C library's malloc as before. Add `--no-mimalloc` and `--mimalloc` options to override the automatic behaviour.

Add CI tests for `--no-mimalloc`. The CI systems have compilers that support atomics, so they now default to mimalloc.

See #1623 for motivation and #1634 for the previous iteration

MasterDuke17 commented:
Huh, looks like the probe doesn't find stdatomic.h on Windows?

On Fri, Jan 7, 2022 at 10:51 AM Nicholas Clark ***@***.***>
wrote:

> Permit building with mimalloc instead of libc malloc
>
> Implement this by statically include mimalloc as a single source. Default
> to using mimalloc if Configure.pl's probe detects that the compiler
> provides C11 atomics, otherwise continue to use the C library's malloc as
> before. Add --no-mimalloc and --mimalloc options to override the
> automatic behaviour.
>
> Add CI tests for --no-mimalloc. The CI systems have compilers that
> support atomics, so they now default to mimalloc.
>
> See #1623 <https://github.com/MoarVM/MoarVM/issues/1623> for motivation
> and #1634 <https://github.com/MoarVM/MoarVM/pull/1634> for the previous
> iteration
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/MoarVM/MoarVM/pull/1638
> Commit Summary
>
>    - c496e88
>    <https://github.com/MoarVM/MoarVM/pull/1638/commits/c496e887c5b0e716ba2037325a6041465bff9e7b>
>    Convert remaining direct calls to `malloc`/`free`/ to the MVM_* wrappers
>    - 5d46f54
>    <https://github.com/MoarVM/MoarVM/pull/1638/commits/5d46f54f73bc2277b5d5b8897a330cd755a1a7d5>
>    Abstract `strdup` and `strndup` as `MVM_strdup`/`MVM_strndup`
>    - c20ed26
>    <https://github.com/MoarVM/MoarVM/pull/1638/commits/c20ed26400737ced99c01bf5c41c7a598be1ae1e>
>    Link with BCrypt for mingw
>    - 894787c
>    <https://github.com/MoarVM/MoarVM/pull/1638/commits/894787c5d86809b93ec80dfd36c5f6dbad323564>
>    In Configure.pl use a '!' flag to --telemeh to implement --no-telemeh
>    - 765ce7e
>    <https://github.com/MoarVM/MoarVM/pull/1638/commits/765ce7e6a6c35b44d0a1b9854efbb4fd9276dc0c>
>    Add a Configure.pl probe for working #include <stdatomic.h>
>    - c477267
>    <https://github.com/MoarVM/MoarVM/pull/1638/commits/c47726770f249f682e593f46f777ee1c3f695a15>
>    Add mimalloc submodule
>    - 672750e
>    <https://github.com/MoarVM/MoarVM/pull/1638/commits/672750e966013982a52dfec377616cc289b785dd>
>    Permit building with mimalloc instead of libc malloc
>
> File Changes
>
> (17 files <https://github.com/MoarVM/MoarVM/pull/1638/files>)
>
>    - *M* .gitmodules
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-fe7afb5c9c916e521401d3fcfb4277d5071798c3baf83baf11d6071742823584>
>    (3)
>    - *A* 3rdparty/mimalloc
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-12a1f285dfd6eb8c98386c783f44c625dfc06db8fdb6f63fba703b1563f824ef>
>    (1)
>    - *M* Configure.pl
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-86fb5002d50fced74864782aa7e5bbba8088d56e997ae0db75cc746644896af4>
>    (44)
>    - *M* LICENSE
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-c693279643b8cd5d248172d9c22cb7cf4ed163a3c98c8a3f69c2717edd3eacb7>
>    (1)
>    - *M* azure-pipelines.yml
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-7915b9b726a397ae7ba6af7b9703633d21c031ebf21682f3ee7e6a4ec52837a5>
>    (12)
>    - *M* build/Makefile.in
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-032892f2b40899c7dd66577777bc7d6c0383ec93c6182321453ed306fdfe393d>
>    (2)
>    - *M* build/config.h.in
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-1b46ce6126d0486e52013da87bf9fb3ab23f89dbeaa93b5269e64a7e3fd719e9>
>    (4)
>    - *M* build/probe.pm
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-de203bc323b50681208f340eb8ae81a96be027c181ac1e1063b029757e4cd758>
>    (63)
>    - *M* build/setup.pm
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-ba6609baf53b3f3d02c18be173bbe1fa52739b47523ae83049e43527df92f95c>
>    (4)
>    - *M* src/6model/bootstrap.c
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-9efa55509f715a5af15ef7a1d95f8599622ea31328540ec25dca380ba8c36a99>
>    (2)
>    - *M* src/core/alloc.h
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-960047475d146c94a3474e4ba39b05d94f69c998440582df4886a19750cde011>
>    (20)
>    - *M* src/disp/program.c
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-13d89c271f2c8d9f58e5d5043c2d91e98b963c1a5ecfda2a3c783587cf179806>
>    (4)
>    - *M* src/moar.c
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-2f58429081863e7f7a6e5fd7c31cc6e40f7717606e45769f3709b9acf75eda03>
>    (4)
>    - *M* src/moar.h
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-7c458c7a4a4f9096d1e38f454cb00ba0748e97a84370e58d56fd087c7f753763>
>    (11)
>    - *M* src/profiler/heapsnapshot.c
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-2c8220ad172a8f7f7357010edc179c3e22529af49d6703c443ed47d4d1bd567f>
>    (2)
>    - *M* src/profiler/telemeh.c
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-a3fdef47716adb53ef443f57bb5c40a93088821f0bf8da1e191d39925a32d738>
>    (4)
>    - *M* tools/release.sh
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-5c05dc4c302e9347cb2e91efd67fe5bf69ab7f93cffe21e8bdf9f99badeb9e4f>
>    (2)
>
> Patch Links:
>
>    - https://github.com/MoarVM/MoarVM/pull/1638.patch
>    - https://github.com/MoarVM/MoarVM/pull/1638.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MoarVM/MoarVM/pull/1638>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOHYUNZP55G2R2BDYUTI6DUU3ARRANCNFSM5LOQTLPQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

MasterDuke17 commented:
Noticed two typos in the description of the new args in Configure.pl.

"we default is to" -> "we default to"

"Specify C<--no-mimalloc> to force use of mimalloc" -> "Specify
C<--mimalloc> to force use of mimalloc"

On Fri, Jan 7, 2022 at 10:51 AM Nicholas Clark ***@***.***>
wrote:

> Permit building with mimalloc instead of libc malloc
>
> Implement this by statically include mimalloc as a single source. Default
> to using mimalloc if Configure.pl's probe detects that the compiler
> provides C11 atomics, otherwise continue to use the C library's malloc as
> before. Add --no-mimalloc and --mimalloc options to override the
> automatic behaviour.
>
> Add CI tests for --no-mimalloc. The CI systems have compilers that
> support atomics, so they now default to mimalloc.
>
> See #1623 <https://github.com/MoarVM/MoarVM/issues/1623> for motivation
> and #1634 <https://github.com/MoarVM/MoarVM/pull/1634> for the previous
> iteration
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/MoarVM/MoarVM/pull/1638
> Commit Summary
>
>    - c496e88
>    <https://github.com/MoarVM/MoarVM/pull/1638/commits/c496e887c5b0e716ba2037325a6041465bff9e7b>
>    Convert remaining direct calls to `malloc`/`free`/ to the MVM_* wrappers
>    - 5d46f54
>    <https://github.com/MoarVM/MoarVM/pull/1638/commits/5d46f54f73bc2277b5d5b8897a330cd755a1a7d5>
>    Abstract `strdup` and `strndup` as `MVM_strdup`/`MVM_strndup`
>    - c20ed26
>    <https://github.com/MoarVM/MoarVM/pull/1638/commits/c20ed26400737ced99c01bf5c41c7a598be1ae1e>
>    Link with BCrypt for mingw
>    - 894787c
>    <https://github.com/MoarVM/MoarVM/pull/1638/commits/894787c5d86809b93ec80dfd36c5f6dbad323564>
>    In Configure.pl use a '!' flag to --telemeh to implement --no-telemeh
>    - 765ce7e
>    <https://github.com/MoarVM/MoarVM/pull/1638/commits/765ce7e6a6c35b44d0a1b9854efbb4fd9276dc0c>
>    Add a Configure.pl probe for working #include <stdatomic.h>
>    - c477267
>    <https://github.com/MoarVM/MoarVM/pull/1638/commits/c47726770f249f682e593f46f777ee1c3f695a15>
>    Add mimalloc submodule
>    - 672750e
>    <https://github.com/MoarVM/MoarVM/pull/1638/commits/672750e966013982a52dfec377616cc289b785dd>
>    Permit building with mimalloc instead of libc malloc
>
> File Changes
>
> (17 files <https://github.com/MoarVM/MoarVM/pull/1638/files>)
>
>    - *M* .gitmodules
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-fe7afb5c9c916e521401d3fcfb4277d5071798c3baf83baf11d6071742823584>
>    (3)
>    - *A* 3rdparty/mimalloc
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-12a1f285dfd6eb8c98386c783f44c625dfc06db8fdb6f63fba703b1563f824ef>
>    (1)
>    - *M* Configure.pl
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-86fb5002d50fced74864782aa7e5bbba8088d56e997ae0db75cc746644896af4>
>    (44)
>    - *M* LICENSE
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-c693279643b8cd5d248172d9c22cb7cf4ed163a3c98c8a3f69c2717edd3eacb7>
>    (1)
>    - *M* azure-pipelines.yml
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-7915b9b726a397ae7ba6af7b9703633d21c031ebf21682f3ee7e6a4ec52837a5>
>    (12)
>    - *M* build/Makefile.in
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-032892f2b40899c7dd66577777bc7d6c0383ec93c6182321453ed306fdfe393d>
>    (2)
>    - *M* build/config.h.in
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-1b46ce6126d0486e52013da87bf9fb3ab23f89dbeaa93b5269e64a7e3fd719e9>
>    (4)
>    - *M* build/probe.pm
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-de203bc323b50681208f340eb8ae81a96be027c181ac1e1063b029757e4cd758>
>    (63)
>    - *M* build/setup.pm
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-ba6609baf53b3f3d02c18be173bbe1fa52739b47523ae83049e43527df92f95c>
>    (4)
>    - *M* src/6model/bootstrap.c
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-9efa55509f715a5af15ef7a1d95f8599622ea31328540ec25dca380ba8c36a99>
>    (2)
>    - *M* src/core/alloc.h
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-960047475d146c94a3474e4ba39b05d94f69c998440582df4886a19750cde011>
>    (20)
>    - *M* src/disp/program.c
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-13d89c271f2c8d9f58e5d5043c2d91e98b963c1a5ecfda2a3c783587cf179806>
>    (4)
>    - *M* src/moar.c
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-2f58429081863e7f7a6e5fd7c31cc6e40f7717606e45769f3709b9acf75eda03>
>    (4)
>    - *M* src/moar.h
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-7c458c7a4a4f9096d1e38f454cb00ba0748e97a84370e58d56fd087c7f753763>
>    (11)
>    - *M* src/profiler/heapsnapshot.c
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-2c8220ad172a8f7f7357010edc179c3e22529af49d6703c443ed47d4d1bd567f>
>    (2)
>    - *M* src/profiler/telemeh.c
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-a3fdef47716adb53ef443f57bb5c40a93088821f0bf8da1e191d39925a32d738>
>    (4)
>    - *M* tools/release.sh
>    <https://github.com/MoarVM/MoarVM/pull/1638/files#diff-5c05dc4c302e9347cb2e91efd67fe5bf69ab7f93cffe21e8bdf9f99badeb9e4f>
>    (2)
>
> Patch Links:
>
>    - https://github.com/MoarVM/MoarVM/pull/1638.patch
>    - https://github.com/MoarVM/MoarVM/pull/1638.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MoarVM/MoarVM/pull/1638>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOHYUNZP55G2R2BDYUTI6DUU3ARRANCNFSM5LOQTLPQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

nwc10 commented:
> Huh, looks like the probe doesn't find stdatomic.h on Windows?

I don't have access to Win32, so the CI run is the first time this code has encountered it. You spotted a problem faster than I did.

So, .../include/mimalloc-atomic.h

```
#if defined(__cplusplus)
// Use C++ atomics
#include <atomic>
#define  _Atomic(tp)            std::atomic<tp>
#define  mi_atomic(name)        std::atomic_##name
#define  mi_memory_order(name)  std::memory_order_##name
#elif defined(_MSC_VER)
// Use MSVC C wrapper for C11 atomics
#define  _Atomic(tp)            tp
#define  ATOMIC_VAR_INIT(x)     x
#define  mi_atomic(name)        mi_atomic_##name
#define  mi_memory_order(name)  mi_memory_order_##name
#else
// Use C11 atomics
#include <stdatomic.h>
#define  mi_atomic(name)        atomic_##name
#define  mi_memory_order(name)  memory_order_##name
#endif
```

Yet again, looks like "there are standards, and there's what MSVC provides". So I guess I need to ignore the probe if the compiler is MSVC, and just "go for it" with mimalloc.

nwc10 commented:
I rebased this, and changed the key in `%Config` from `mimalloc` to `use_mimalloc`

I thought that this key is clearer (long term), and I thought it cleaner to have a revisionist history where it was "correct" from the first commit, rather than an extra commit to have to read and think about.

I haven't dug into the CI tests - need to go AFK.

MasterDuke17 commented on Configure.pl:
```diff
@@ -549,6 +548,37 @@ sub uniq {
 build::probe::ptr_size(\%config, \%defaults);
+build::probe::stdatomic(\%config, \%defaults);
+
+$config{mimalloc} = $args{mimalloc};
```
Looks like a mixture of `mimalloc` and `use_mimalloc` might be the problem.

MasterDuke17 commented:
> I rebased this, and changed the key in %Config from mimalloc to use_mimalloc

I don't care strongly either way, but to play devil's advocate, we have existing `dtrace` and `telemeh` flags (i.e., not `use_dtrace` and `use_telemeh`).

MasterDuke17 commented on src/core/alloc.h:
```diff
@@ -26,7 +38,11 @@ MVM_STATIC_INLINE void * MVM_realloc(void *p, size_t size) {
 
 MVM_STATIC_INLINE void * MVM_recalloc(void *p, size_t old_size, size_t size) {
+#ifdef MVM_USE_MIMALLOC
+    void *ptr = mi_realloc(p, size);
```
There is an `mi_recalloc` https://github.com/microsoft/mimalloc/blob/master/include/mimalloc.h#L227, but it expects a new count and size (like regular calloc), not old size and new size. There are only a couple uses of `MVM_recalloc`, maybe a followup commit/PR could switch `MVM_recalloc` to use `mi_recalloc` and fix all the uses.

MasterDuke17 replied:
https://gist.github.com/MasterDuke17/24f85769c96d78166138d0c18c792020 has a diff against master that I believe changes all the uses of `MVM_recalloc` and just needs a tweak in the implementation of `MVM_recalloc` to use `mi_recalloc`.

MasterDuke17 commented on Configure.pl:
```diff
@@ -549,6 +548,37 @@ sub uniq {
+build::probe::stdatomic(\%config, \%defaults);
+
+$config{use_mimalloc} = $args{mimalloc};
+if (!defined $config{mimalloc}) {
```
Whoops, like one was missed.

nwc10 replied:
Bother yes, well spotted. Clearly the order in https://colabti.org/irclogger/irclogger_log/moarvm?date=2022-01-09#l62 was wrong. I guess next time it ought be

1. make coffee
2. re-read commit again
3. only then (if good) push

nwc10 commented:
> I don't care strongly either way, but to play devil's advocate, we have existing `dtrace` and `telemeh` flags (i.e., not `use_dtrace` and `use_telemeh`).

Or `--use-dtrace` ...

The context was that I was working further on what became #1640 and realised that in `build/config.h.in` (nearly) everything was consistent *verb_noun*, where *verb* was pretty descriptive - basically *can* or *has*. So "mimalloc" is "yes, right, what about it?" We have it? (always true - we ship it) We can use it? (no, not necessarily true if we could but chose to disable it) So this configuration variable was "we have enabled it" hence "use". Readability seemed important here.

I didn't change the comand-line flag - which (somewhat) then increases inconsistency. Except that it would want to be `--use-mimalloc` not `--use_mimalloc` and I don't think that our configuration keys or (C macros) can use `-` in place of `_`. But I somewhat feel that the command line flags are already a lost cause.

MasterDuke17 commented:
> I didn't change the comand-line flag - which (somewhat) then increases inconsistency. Except that it would want to be `--use-mimalloc` not `--use_mimalloc` and I don't think that our configuration keys or (C macros) can use `-` in place of `_`. But I somewhat feel that the command line flags are already a lost cause.

I wouldn't object to changing `--dtrace`, `--telemeh`, and `--mimalloc` to `--use-*`.
